dataplane/memory_manager.h

```diff
@@ -42,6 +42,44 @@ class memory_manager
 	        const uint64_t size,
 	        const std::function<void(void*)>& destructor = [](void*) {});
 
+	class Deleter
+	{
+		memory_manager* manager_;
+
+	public:
+		Deleter(memory_manager* const manager) :
+		        manager_{manager} {}
+		Deleter(Deleter&& other) = default;
+		Deleter(const Deleter& other) = default;
+		Deleter& operator=(Deleter&& other) = default;
+		Deleter& operator=(const Deleter& other) = default;
+		template<typename T>
+		void operator()(T* ptr)
+		{
+			manager_->destroy(ptr);
+		}
+	};
+
+	template<typename T>
+	using unique_ptr = std::unique_ptr<T, Deleter>;
+
+	template<typename type,
+	         typename... args_t>
+	unique_ptr<type> create_unique(const char* name,
+	                               const tSocketId socket_id,
+	                               const uint64_t size,
+	                               const args_t&... args)
+	{
+		return std::unique_ptr<type, Deleter>{
+		        reinterpret_cast<type*>(alloc(name,
+		                                      socket_id,
+		                                      size,
+		                                      [](void* pointer) {
+			                                      reinterpret_cast<type*>(pointer)->~type();
+		                                      })),
+		        Deleter{this}};
+	}
```

Comment:
I'm not sure about that.. I think we can just use custom deleter directly with `std::unique_ptr` and have the same functionality by using simpler constructs.

```cpp
template<typename T, typename... Args>
std::unique_ptr<T, std::function<void(T*)>> create_unique(const char* name,
                                                          tSocketId socket_id,
                                                          size_t size,
                                                          Args&&... args)
{
	T* raw_pointer = alloc(name, socket_id, size, [](void* ptr) { reinterpret_cast<T*>(ptr)->~T(); });
	return std::unique_ptr{raw_pointer, [this](T* ptr) { this->destroy(ptr); }};

}
```

Reply:
Can't seem to see how using `std::function` results in simpler constructs here. Except for moving raw_pointer allocation out of return statement those are virtually identical.

On the other hand using `std::function` means that a deleter has to store not only pointer to `memory_manager` but also a pointer for vtabe. This would double deleter's memory footprint.

Reply:
Yeah, you right. The overhead of `std::function` is indeed larger than that of a custom deleter class.
let's just move raw pointer allocation out then and do other minor corrections:
- use typename `T, Args` instead of `type, args_t`
- remove unnecessary const
- `Args&&` instead of `const args_t&`



Ehh.. we could speculate a little about using `auto` as a return type auto so the compiler will use the lambda directly as a callable object, and a std::function will not be created.
But since we need to store that unique_ptr as a field in another class, we will need the type. So we can extract lambda to a function with `auto` as a return type and then use `std::result_of_t`. Smth like this:
```cpp
    auto CreateDeleter() {
        return [this](auto* ptr) {
            this->destroy(ptr);
        };
    }
   using Deleter = std::result_of_t<decltype(&memory_manager::CreateDeleter)>;

    template<typename T, typename... Args>
    auto create_unique(const char* name, tSocketId socket_id, size_t size, Args&&... args) {
        T* raw_pointer = reinterpret_cast<T*>(alloc(name, socket_id, size, [](void* ptr) {
            reinterpret_cast<T*>(ptr)->~T();
        }));
        return std::unique_ptr(raw_pointer, CreateDeleter());
    }
    //...
    using Deleter = std::result_of_t<decltype(&memory_manager::CreateDeleter)>;
```

But I don't think that this is better than just a custom class-deleter.

Reply:
oh, we can do that even without a function, just call `::deleter_type` from `unique_ptr`. Didn't know that